### DISCUSSION TITLE
Is it possible to pass stdin to a sidecar binary?

### DISCUSSION THREAD

**Discussion by salmenf** [tauri-apps/tauri#4440] [Q&A] [answered]
Hi, I'm using Tauri's sidecar feature to bundle an esbuild binary. This works well so far and with the [`Command.sidecar` API](https://tauri.app/v1/api/js/classes/shell.command/#sidecar), I'm able to use it.
However, one issue arose: esbuild offers a [transform API](https://esbuild.github.io/api/#transform-api) to create a bundle from a JS string, looking like this:
```
echo 'let x: number = 1' | esbuild --loader=ts
```
Since I can only specify args with the `Command.sidecar` API, I'm not sure how to replicate this example via Tauri. Unfortunately, esbuild offers no CLI option to pass the input string as an argument. Any ideas?

**Comment by FabianLars** [comment]
Does this help you? https://github.com/tauri-apps/tauri/discussions/2984#discussioncomment-1716861

**Reply by salmenf** [reply]
Thanks for the quick reply! I tried that, now I'm having some trouble accessing the output. This is what I'm running as an experiment:
```ts
const command = Command.sidecar("../binaries/esbuild", "--loader=ts")
command.stdout.on("data", data => console.log(data))
command.stderr.on("data", data => console.log(data))
command.on("error", data => console.log(data))
command.on("close", data => console.log(data))
const child = await command.spawn()
console.log(child.pid)
await child.write("echo 'let x: number = 1'")
```

Only the PID is logged, the event listeners do nothing as far as I can tell. I would expect either the result from stdout or an error from stderr.

**Reply by FabianLars** [reply]
hmm tricky, a common issue can be that esbuild's output is in a weird format and/or without a good line delimiter. Stuff like this is pretty hard to debug :/

**Reply by salmenf** [reply]
I see! For now, I implemented a workaround specific to `esbuild` using their normal `build` API instead of `transform`. Thanks anyway for your help.

**Reply by kennethnym** [reply]
Hi @salmenf, do you have any code snippet for how you use the build api? I tried to follow their docs and came up with this:
```js
const esbuild = Command.sidecar(SIDECAR.esbuild, [
  "main.js",
  "--bundle",
  "--minify",
  "--platform=node",
  "--format=cjs",
  "--outfile=dist.js"
], {
  cwd: tempDir,
  env: {
    NODE_PATH: await resourceDir()
  }
})

esbuild
  .stderr
  .on("data", (err) => {
    console.log("err", err)
  })
  esbuild
  .stdout
  .on("data", (data) => {
    console.log("stdout", data)
  })

esbuild
.on("error", (err) => {
  console.log("err", err)
})
esbuild
.on("close", async () => {
  const text = await readTextFile(await join(tempDir, "main.js"))
  console.log("text", text)
})

await esbuild.spawn()
```
However, the promise will never resolve, nothing gets logged, and the file (`dist.js`) is never generated. I have no idea why nothing is happening.

Edit: I have tried running the binary directly in the terminal and everything works just fine, so there must be something wrong in between the binary and Tauri.

**Reply by salmenf** [reply]
Hi @kennethnym! I took a quick look at my own code. Mine is a bit convoluted, but it's in use and working. Not sure why yours isn't working, to be honest. The main difference is that I use use `execute` and `await`, then handle the output, instead of `spawn` and `on`. Hope that helps a bit. :)

```ts
export async function bundle(args: string[] = []) {
  const output = await Command.sidecar("../../../binaries/esbuild", args).execute()
  if(output.code !== 0) {
    return Error(output.stderr)
  }
  else {
    return {data: output.stdout}
  }
}

// This is just some call from my code, the variables are not important
bundle([`${entrypointPath}`, "--bundle", `--outfile=${bundlePath}.js`, `--format=esm`])
```

**Reply by kennethnym** [reply]
Hi, appreciate your response! I also tried calling `execute` instead of `spawn`, but it would just hang and never resolve, presumably because the `"close"` event never happened in the first place. How did you get your esbuild binary? Was it copied from `node_modules/esbuild/bin/esbuild` (obtained by doing `npm i -D esbuild`)?

**Reply by salmenf** [reply]
No worries! I use the specific package for each target platform I have, those packages directly contain the executable with the correct architecture. For example, I use `esbuild-windows-64`, pull the executable from that package and rename it with the Tauri/Rust naming convention (using a small script) so that it is used as a sidecar when building for that platform. There, `esbuild-windows-64` becomes `esbuild-x86_64-pc-windows-msvc.exe`.

**Comment by lanc33llis** [comment, marked as answer]
For those coming across this issue and have no other options except stdin (e.g. external binaries maintained by others like in my use-case), the issue is that stdin is not being closed. It expects more input which of course never leads to any stdout. I'm not sure how to do this on the client-side (JS), but you can do it easily in Rust with ```drop``` then invoke the command on the frontend. 

Backend
```rust
#[tauri::command]
async fn command_to_spawn_sidecar(text: &str) -> Result<String, String> {
  let (mut rx, mut child) = Command::new_sidecar("my-sidecar")
    .expect("failed to create `my-sidecar` binary command")
    .spawn()
    .expect("Failed to spawn sidecar");

  child.write(text.as_bytes()).expect("Failed to write to sidecar");
  drop(child);

  let mut output = String::new();
  while let Some(event) = rx.recv().await {
    if let CommandEvent::Stdout(line) = event {
      output.push_str(&line);
    }
  }

  Ok(output)
}
```

Frontend React Hook
```tsx
import useTauriApi from "../useTauriApi";
// hook to invoke Tauri api

const useFunctionApi: (
  text: string
) => [response: string, loading: boolean] = (text: string) => {
  const [response, loading] = useTauriApi<string>("command_to_spawn_sidecar", {
    text,
  });

  return [response, loading];
};

export default useFunctionApi;
```

Usage
```tsx
import useFunctionApi from "../../../hooks/api/useFunctionApi";

const ExampleComp = ({ text }: { text: string }) => {
  const [output, loading] = useFunctionApi(text);

  return <div>{output}</div>;
};

export default ExampleComp;

```

**Reply by salmenf** [reply]
Interesting! Thanks for your solution, I was still looking for something like that.

@FabianLars Maybe the Tauri API could be be extended with some `child.closeStdin` method, or even just a flag for the `child.write` method like `closeAfter=true`? Judging from @lanc33llis code, it's quick to implement. I don't know any Rust, unfortunately, but I can write all of this up as an issue if that helps.

**Reply by FabianLars** [reply]
@salmenf Can you open a feature request for this please? 🙏
